Initial support for passing ssh options from cli.

RHEmployee opened the pull request.

lukaszachy commented on tmt/steps/provision/connect.py:
```diff
@@ -108,6 +109,13 @@ def go(self):
             data['key'] = key
 
+        if ssh_option:
+            self.info('ssh options', ssh_option, 'green')
```
Should be formatted to be more user friendly. `run --ssh-option StrictHostKeyChecking=no` prints 
`ssh options: ('StrictHostKeyChecking=no',)`

Also could it be printed under same name as  cli option (`--ssh-option`)?

RHEmployee replied:
Hi. Good point. However, there can be more --ssh-option on cli.
--ssh-option blah=1 --ssh-option kuku=2

Maybe it can by printed line by line?

ssh-option: blah=1
ssh-option: kuku=2
...

Line by line approach seems to be nice from my pow.

lukaszachy replied:
Problem is that values aren't ready for copy&paste if there are multiple values (comma separated, some values can be omitted and last value has "and" before it)
So IMHO pointing user to the correct option is enough.

lukaszachy reviewed (changes requested):
Will you change also provision/testcloud ?

RHEmployee commented:
Ok. I changed how ssh-option is rendered to logs.
Not cut and pasta, but at least something.

psss commented:
Summary from the last week's discussion:

* Update the default ssh options as proposed in the issue
* Provide the command line only way to override
* Document this new tmt default somewhere

Example usage: 

    tmt run provision --how connect --ssh-option=OPTION

RHEmployee commented:
> Summary from the last week's discussion:
> 
>     * Update the default ssh options as proposed in the issue
Done
> 
>     * Provide the command line only way to override
Done - maybe not intuitive but works
> 
>     * Document this new tmt default somewhere
Where should I document this issue? Where is somewhere? Official docs? Where I can grab it?
> 
> 
> Example usage:
> 
> ```
> tmt run provision --how connect --ssh-option=OPTION
Should be done.
> ```

Thanks!

lukaszachy commented:
> > ```
> > * Document this new tmt default somewhere
> > ```
> 
> Where should I document this issue? Where is somewhere? Official docs? Where I can grab it?

Official docs are at readthedocs and are generated from this repo, https://tmt.readthedocs.io/en/stable/contribute.html#docs mentions details.

@psss Any idea how this option can be done only to GuestSSH-like provisioners? For local, podman it doesn't make sense.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -310,6 +315,17 @@ def _ssh_options(self, join=False):
+                options.insert(index, ssh_op)
+                index += 1
+        except TypeError:
+            pass
```
* I don't understand what is `except TypeError` guarding against. What kind of issue may lead to `TypeError` being raised? 
* the `index += 1` smells like a job for `enumerate()`, but... IIUIC, loop adds new options to `options`, at the beginning. Wouldn't `options = ssh_options + options` do the exactly the same job?

Seems to me this could be simplified, given the default value of `self.ssh_option` to be `[]` AFAICT:

```python
# ssh cmdline grabs first cli option and ignores others.
# This way ve can override ssh default ssh options.
options = [v for opt in self.ssh_option for v in ("-o", opt)] + options

# Or, since f-strings are a thing and nested comprehensions harder to follow:
options = [f'-o{opt}' for opt in self.ssh_option] + options
```

* no-op when `self.ssh_option` is empty,
* no need for `insert()` or extra variables,
* follows the `-oFoo` approach already stored in `options`.

RHEmployee replied:
Thanks. Changed.

RHEmployee commented:
> Any idea how this option can be done only to GuestSSH-like provisioners? For local, podman it doesn't make sense.

Only connection plugin is supported now. Otherwise, --ssh-option (tested on virtual case) is ignored.
Because you need to update plugins to handle all cli options and only connection plugin is modified this way.

lukaszachy commented:
> > Any idea how this option can be done only to GuestSSH-like provisioners? For local, podman it doesn't make sense.
> 
> Only connection plugin is supported now. Otherwise, --ssh-option (tested on virtual case) is ignored. Because you need to update plugins to handle all cli options and only connection plugin is modified this way.

So it is even worse, because help prints this option for each provision plugin.

psss commented:
> So it is even worse, because help prints this option for each provision plugin.

Just quick brainstorm: Would it make sense to apply the changes to the new `GuestSSH` class proposed in #1194?

happz commented:
> > So it is even worse, because help prints this option for each provision plugin.
> 
> Just quick brainstorm: Would it make sense to apply the changes to the new `GuestSSH` class proposed in #1194?

@psss 1. absolutely, `GuestSSH` is the right place for it. 2. I'd be happy with both outcomes, i.e. merging this PR and I'd refactor my patch to put this new functionality to its right place, or merge my patch first and I can help with refactoring and review of this patch after that. Both ways are fine with me.

psss commented:
> or merge my patch first and I can help with refactoring and review of this patch after that

Let's go this way. The guest split change has just two minor comments, let's finish that one first.

psss commented:
> Where should I document this issue? Where is somewhere? Official docs? Where I can grab it?

I guess a good place could be under the [provision spec](https://tmt.readthedocs.io/en/latest/spec/plans.html#provision) section?

> So it is even worse, because help prints this option for each provision plugin.

Yes, we definitely do not want this ssh-related option for  `provision` plugins where no `ssh` is involved.

> @psss Any idea how this option can be done only to GuestSSH-like provisioners?

I've rebased on the latest `main`, so the new `GuestSsh` class is now available. What about something like this:

 * Each `ProvisionPlugin` class could have a class attribute `_guest_class` pointing to the appropriate guest implementation
 * Each `Guest` class would provide an `options()` method generating `click` options specific to it
 * `GuestSsh` would define the new `--ssh-config` option in its `options()` method
 * During `ProvisionPlugin.options()` relevant guest class options would be gathered as well

@happz, @lukaszachy, @thrix, what do you think?

Anyway, this needs more discussion/time to finish, let's not block `1.14` on this.

happz commented:
> * Each `ProvisionPlugin` class could have a class attribute `_guest_class` pointing to the appropriate guest implementation
> * Each `Guest` class would provide an `options()` method generating `click` options specific to it
> * `GuestSsh` would define the new `--ssh-config` option in its `options()` method
> * During `ProvisionPlugin.options()` relevant guest class options would be gathered as well
> 
> @happz, @lukaszachy, @thrix, what do you think?
> 
> Anyway, this needs more discussion/time to finish, let's not block `1.14` on this.

Sounds good to me, +1.

psss commented:
Ok, let's go this way then.

psss commented:
@RHEmployee, are you ok with the [proposed](https://github.com/teemtee/tmt/pull/1088#issuecomment-1147476632) approach?

thrix commented:
@psss @lukaszachy @happz I would also think about if we could set at least some better defaults for `ServerAliveInterval` and `ServerAliveCountMax` right away?

thrix commented:
@psss so I made some experiments and:

* seems setting `ServerAliveInterval` does not anyhow affect reboot, client nicely terminates
* once I set `ServerAliveInterval` the time client gives up on unresponsive server is deterministic - `ServerAliveInterval` * `ServerAliveCountMax`
* without setting `ServerAliveInterval` ssh client can hang for very long time before giving up (it is not deterministic)

So to mitigate the original reported issue by default, so user does not need to care about it, I would set the maximum time to 5 minutes, the same as the default test duration is, and document it.

So:

```
ServerAliveInterval = 60
ServerAliveCountMax = 5
```

Would be what I would go with, 5 minutes, same as default test duration.

RHEmployee commented:
Hi. Rebased and added reasonable defaults.

However:
https://github.com/teemtee/tmt/pull/1088#issuecomment-1147520492
do you think this should land is this commit? Maybe we should open new request for this?

lukaszachy commented:
> Hi. Rebased and added reasonable defaults.
> 
> However: [#1088 (comment)](https://github.com/teemtee/tmt/pull/1088#issuecomment-1147520492) do you think this should land is this commit? Maybe we should open new request for this?

In this commit please - This pr adds ssh options to _all_ provision methods (and e.g. in local or podman in doesn't make sense)

psss commented:
Originally I thought that the [proposed](https://github.com/teemtee/tmt/pull/1088#issuecomment-1147476632) solution will be covered together with the ssh options in this PR as they are very closely related, but we can do a separate PR as well if you see any benefit in that approach.

RHEmployee commented:
From my POV, different MR would be building block for wider options handling solution which is not tied only with ssh options.
I'm open to discussion however.

psss commented:
> From my POV, different MR would be building block for wider options handling solution which is not tied only with ssh options.

Ok, let's do a separate pull request for that. @RHEmployee, are you going to work on that?

RHEmployee commented:
> > From my POV, different MR would be building block for wider options handling solution which is not tied only with ssh options.
> 
> Ok, let's do a separate pull request for that. @RHEmployee, are you going to work on that?

Hi. Sorry, missed some notifications from github (some filter issues :().
I talked about this approach with @lukaszachy and he told me, that maybe he can do it.
However, it would be nice if I could take a look. Not sure how to implement but I'm eager to find out how :).

thrix commented:
@lukaszachy so are you looking into that next PR?

lukaszachy commented:
> @lukaszachy so are you looking into that next PR?

No. All I said on irc was that it would be better if some of core tmt devels creates the PR.

RHEmployee commented:
Ok. I started:
https://github.com/teemtee/tmt/pull/1385

It's heavily work in progress but it's start.

However, I'm not sure about passing:
During ProvisionPlugin.options() relevant guest class options would be gathered as well

Not sure where ProvisionPlugin.options()  is called.

psss commented:
> However, I'm not sure about passing:
> During ProvisionPlugin.options() relevant guest class options would be gathered as well

Well, each respective plugin usually calls it during its options definition: Here's an example:

https://github.com/teemtee/tmt/blob/0ef074a140cf1d55403666b2228a33bb6cb8ab5a/tmt/steps/provision/testcloud.py#L230-L233

RHEmployee commented:
Hi. Seems to be ready for review.

happz commented on tmt/steps/provision/artemis.py:
```diff
@@ -347,7 +350,7 @@ def options(cls, how: Any = None) -> List[tmt.options.ClickOptionDecoratorType]:
                      f'{DEFAULT_RETRY_BACKOFF_FACTOR} by default.',
                 ),
-            ]) + super().options(how)
+            ]) + super().options(how) + cls._guest_class.options()
```
* I'd definitely add `_guest_class` attribute to the base class (`tmt.steps.provision.ProvisionPlugin`) to begin with, even without a value because pretty much every derived plugin provides its own class (except `connect` which is perfectly happy with the default `GuestSsh`). So I'd start with `_guest_class: Type[Guest] = Guest` in `ProvisionPlugin`.
* `GuestSsh` is definitely not the correct class - note that all plugins except `connect` have their own subclass of `Guest`. In this case, it should definitely be `GuestArtemis` (and `GuestTestcloud` in `testcloud.py`, and so on).

To build on top of that, I'd suggest a few more extra steps:

* add `options()` method to `Guest` class, even if it'd be returning an empty list.
* add `options()` method to `GuestSsh`, to return `super().options()` plus the new `--ssh-option`
* add custom `options()` method to `ProvisionPlugin` class, to return `super().options()` plus `self._guest_class.options()`

After this, you wouldn't need to spike every plugin with extra `+ cls._guest_class.options()`, because:

* all plugins would keep calling `super().options(how)` as they do now, nothing new there.
* `PreparePlugin.options()` would be clalled (thanks to `super()`), and return its own options (probably none at this point) plus those it gets from `self._guest_class.options()` - this would be new.
* all plugins using SSH-reachable guest classes derive their guest classes from `GuestSsh` and therefore would get `--ssh-option` automatically: `ProvisionArtemis.options()` calls `ProvisionPlugin.options()` which calls `self._guest_class.options()`, which is `GuestArtemis.options()` and that's `GuestSsh.options()`. The dummy `Guest.options()` returning an empty list would make it work with guests and plugins that don't use SSH.

RHEmployee replied:
Hi. Thanks for heads up and review.
Should be done.
Please take a look.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -42,233 +42,6 @@ class CheckRsyncOutcome(enum.Enum):
     INSTALLED = 'installed'
 
 
-class ProvisionPlugin(tmt.steps.GuestlessPlugin):
```
This looks like a failed rebase, I certainly would not expect so much code to be moved around. `ProvisionPlugin` class has been recently moved in front of `Provision` class, because `Provision._base_plugin_class` points to `ProvisionPlugin`.

RHEmployee replied:
Hm.. I moved Guest class on front because I need the type Guest to be known in class ProvisionPlugin(tmt.steps.GuestlessPlugin).

happz replied:
Aaah, that's it, thanks.

happz commented on tmt/steps/provision/artemis.py:
```diff
@@ -279,6 +279,9 @@ class ProvisionArtemis(tmt.steps.provision.ProvisionPlugin):
     _guest = None
 
+    # Provide additional options
+    _guest_class = tmt.steps.provision.GuestSsh
```
`_guest_class` should point to `ArtemisGuest`.

This applies to other modules, too: if they have their own guest class, the plugin's `_guest_class` must point to this class.

RHEmployee replied:
Ok. But I'm not sure how to handle this without moving ArtemisGuest class on top of ProvisionArtemis because there is already:

NameError: name 'GuestArtemis' is not defined

erorr

psss replied:
Yes, split into two commits helped a lot.

RHEmployee replied:
Ok. Thanks for review.

RHEmployee replied:
NOTE: just before split to more commits.

psss replied:
Ok, let me know, when the two commits are ready for review.

RHEmployee replied:
Thanks. Any ideas how to rebase to change #1654?
So split is now not needed if I understand it correctly.

happz replied:
A rebase on top of the branch in #1654 should do the trick.

RHEmployee replied:
Rebased on top of https://github.com/teemtee/tmt/pull/1654

happz replied:
And also modify this MR, to use `reorder-guests-and-plugins` branch as its target, Github should than show just the changes you made.

RHEmployee commented:
It's ready for review.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -454,6 +459,241 @@ def requires(cls) -> List[str]:
         return []
 
 
+class ProvisionPlugin(tmt.steps.GuestlessPlugin):
```
I don't understand the move of `ProvisionPlugin` and `Provision` classes, I thought I moved them to the necessary place in https://github.com/teemtee/tmt/pull/1439. The order should have been guest classes -> plugin class -> step class. Am I missing something, is there a reason to move these two classes in between `Guest` and `GuestSsh` classes?

psss replied:
I believe commit 1b9e4367787960ffeac0804177fd31a7f1ec302a should be dropped. The class reordering in #1654 has been merged, so you can rebase on the latest `main`.

happz commented on tmt/steps/provision/artemis.py:
```diff
@@ -508,9 +511,12 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
                 ),
-            ] + super().options(how)
+            ]
+        return options
```
Changes to `options()` seem to be not needed, actually undoing changes made by previous patches.

happz commented on tmt/steps/provision/artemis.py:
```diff
@@ -508,9 +511,12 @@ def options(cls, how: Optional[str] = None) -> List[tmt.options.ClickOptionDecor
-    # FIXME: ignore - https://github.com/teemtee/tmt/issues/1437
+    # More specific type is a violation of Liskov substitution principle, and mypy
+    # complains about it - rightfully so. Ignoring the issue which should be resolved
+    # with https://github.com/teemtee/tmt/pull/1439.
```
The `FIXME` comment was correct, probably changed recently during the post-annotations review.

happz commented on tmt/steps/provision/connect.py:
```diff
@@ -127,6 +130,14 @@ def go(self) -> None:
+            for option in ssh_option:
+                self.info('ssh option', option, 'green')
+            self.debug('ssh option provided.')
+            data['ssh_option'] = ssh_option
```
1. `data` is not a dictionary, but an instance of `ConnectGuestData` which is a subclass of `GuestSshData`.
2. `GuestSshData` lacks `ssh_option` field.

happz replied:
3. two other plugins do ignore `--ssh-options`, is that correct?

RHEmployee replied:
3. which two? 
Connection uses ssh, artemis uses ssh, minute tip uses ssh, podman uses something different.
I'm not sure about local plugin.

RHEmployee replied:
1. & 2. not sure how to solve it now. Looks like there were some changes in codebase during development of this change.

happz replied:
WRT 1. `data.ssh_option = ssh_option` should be the flavor of the week.

WRT 2. you need to add `ssh_option` field to `provision.GuestSshData` class, probably as `ssh_option: List[str] = dataclasses.field(default_class=list)`. All SSH-backed guests should then inherit it, with a nice empty list as a default value. We could even do `data.ssh_option = self.get('ssh-option')`, you can then check and/or iterate over `data.ssh_option`, no need for extra variable. And the same one to `GuestSsh` class itself, but with not default value.

WRT 3., `testcloud` and `artemis` - both do now have the `--ssh-option` option, and `testcloud` uses it thanks to        
 magic in `data = TestcloudGuestData(**{key: self.get(key) for key in TestcloudGuestData.keys()})`, but `artemis` plugin ignores it. I suppose it needs a similar bit to the one you added to the `connect` plugin right here.

WRT logging, is it needed? Guest data will be logged later anyway, and we would need to repeate this particular bit in all SSH-backed plugins to stay consistent (right now, `connect` uses `--ssh-option` and logs its content, `testcloud` uses it but logs nothing, `artemis` ignores it completely). So we need something as simple as possible, I'd start with `data.ssh_option = self.get('ssh-option')` with no logging involved, and we would then  - next patch? - add some shared solution, transparent for all plugins, like something happening in `Guest.__init__` or similar.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -519,11 +759,18 @@ def _ssh_socket(self) -> str:
             '-oUserKnownHostsFile=/dev/null',
+            '-oServerAliveInterval=60',
+            '-oServerAliveCountMax=5',
             ]
```
IIUIC, this is where we would `+ self.ssh_option`, to propagate options given on the command line (command-line => `data.ssh_option` => `guest.ssh_option`).

psss commented:
Seems that merging the target branch closes pull requests if created from fork instead of redirecting to `main`. Reopening manually.

happz commented:
Let me try pushing this a bit further. If you don't mind, @RHEmployee, I'd make a couple of changes.

RHEmployee commented:
> Let me try pushing this a bit further. If you don't mind, @RHEmployee, I'd make a couple of changes.

Ok. Thanks.

psss commented:
Obsoleted by #1673.
